### ISSUE TITLE
Keyring test suite

### ISSUE THREAD

**Issue by quixotique** [servalproject/serval-dna#44]
The keyring commands need to be covered by a test suite: **tests/keyring**.

Of particular importance is to ensure that all the code dealing with PINs (keyring-wide and per-identity) is well covered and tested.  This is necessary before we can start using a PIN-locked keyring file to store Serval Project authorities, such as the Rhizome Secret for publishing APK updates (see servalproject/batphone#33).

The `keyring_seed()` function is called whenever the server is started in overlay mode, which creates a PIN-less identity if none is present, so this is already extensively tested.

In order to test other keyring functions, it may be necessary to add more keyring commands or add extra arguments to existing commands.

**Comment by quixotique** [comment]
On 5 Feb 2013, **Will Hawkins** wrote in an email:

In one of our uses of servald we found it would be useful to know if a user-supplied keyring file is valid. The keyring code seems incredibly robust, which is really great. I just wanted to make sure that I understood what the return values are from './servald keyring list':

non-0: Something is obviously wrong with accessing the keyring file in the instance path. For example, there are is no R privilege of the user to that directory.

0: Everything is "okay". This "okay" includes the possibility that the file was empty or smaller than keyring-page-size. In these two cases, and probably others, the keyring code simply truncates that file and creates a new, valid default file. This "okay" also includes the possibility that there is bogus information in the file that does not translate into keys but is simply skipped over when iterating through the keyring.

Does this seem like a reasonable overview? The bottom line is that it seems like your code is robust to the point where, when given a bad keyring file, the serval daemon just works.

Is there any reason why a function that iterates through the keyring and checks for the validity of each key (even if it is pin-protected) would be useful? If so, I'd be happy to write it, but I don't want to do something that people think is obviously a bad idea.

Thanks for all the work that you are all doing on this project. I look forward to talking to you

**Comment by quixotique** [comment]
On 5 Feb 2013, I replied to Will Hawkins:

It is impossible to write a function to check the validity of PIN-protected keys without possessing all the PINs concerned.  Valid entries cannot be distinguished from random data without possessing the keyring PIN and entry PIN.  That is our deliberate design, in order to provide absolute deniability of the existence of identities on a device.

The only validity check that can be performed on the keyring file in the absence of any PINs is to check the file size.  If the keyring PIN is empty, then all the PIN-less entries can also be listed.  So the scope for validation of a keyring file is very limited.

It is also worth noting that the current behaviour of the server (in overlay mode) -- to automatically create a single, PIN-less identity if none is found -- will probably change once we implement full identity management features in Batphone.  It may also be controlled by configuration options that do not yet exist.

**Comment by quixotique** [comment]
The new keyring tests have already revealed a SEGV bug which has been fixed in 705f414a4de47af584fe807502686150bae0f470 and b7185a294fd930c4bc68345c375f52416845730c.

Keyring code inspection arising from the testing has also given rise to issue #57.

**Comment by quixotique** [comment]
This issue has progressed enough:
- There is now a keyring test suite with seven test cases.
- The keyring PIN and entry PINs are tested in isolation and when combined.
- Backward compatibility of the keyring file format is enforced by a test case that dumps a legacy keyring file created by the old code.
- The keyring file pack/unpack code has been rewritten to use an extensible static array declaring each key type, instead of large `switch` statements in separate functions that must be kept in sync.
- The rotated buffer code has been rewritten to use the new `rotbuf` primitives instead of modulo arithmetic, allowing proper detection and handling of buffer overfow cases, to reduce potential bugs/exploits.
- A new `keyring dump` command has been implemented, to assist with testing.
- Merged into **development** in 4e48725c93b46546adc0b8de1dbf75bf3cf433fc

Closing.
